pkg/boot/survey/survey.go

```diff
@@ -0,0 +1,367 @@
+				surv.err.Store(err)
+			} else {
+				surv.err.Store(fmt.Errorf("closed"))
+			}
+			lconn.Close()
+			dconn.Close()
+			return
+		}
+	}()
+
+	return surv, nil
+}
+
+func (surv *Survey) Close() {
+	if err := surv.err.Load(); err == nil {
+		surv.c.Close()
+		for surv.err.Load() == nil {
+		}
+	}
+}
+
+func (surv *Survey) multicastHandler(n int, src net.Addr, buffer []byte) {
+	msg, err := capnp.UnmarshalPacked(buffer[:n])
+	if err != nil {
+		return
+	}
+
+	root, err := cpSurvey.ReadRootSurveyPacket(msg)
+	if err != nil {
+		return
+	}
+
+	switch root.Which() {
+	case cpSurvey.SurveyPacket_Which_request:
+		request, err := root.Request()
+		if err != nil {
+			return
+		}
+		surv.handleRequest(request)
+	case cpSurvey.SurveyPacket_Which_response:
+		response, err := root.Response()
+		if err != nil {
+			return
+		}
+		surv.handleResponse(response)
+	default:
+	}
+}
+
+func (surv *Survey) handleRequest(request cpSurvey.SurveyRequest) {
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	// validate requester
+	envelope, err := request.Src()
+	if err != nil {
+		return
+	}
+
+	var rec peer.PeerRecord
+	if _, err = record.ConsumeTypedEnvelope(envelope, &rec); err != nil {
+		return
+	}
+	if rec.PeerID == surv.h.ID() {
+		return // request comes from itself
+	}
+
+	if dist([]byte(surv.h.ID()), []byte(rec.PeerID))>>uint32(request.Distance()) != 0 {
+		return
+	}
+
+	ns, err := request.Namespace()
+	if err != nil {
+		return
+	}
+
+	if _, ok := surv.mustAdvertise[ns]; !ok {
+		return
+	}
+
+	response, err := surv.buildResponse(ns)
+	if err == nil {
+		go surv.c.Send(response)
+	}
+}
+
+func (surv *Survey) buildResponse(ns string) ([]byte, error) {
+	_, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
+	if err != nil {
+		panic(err)
+	}
+
+	root, err := cpSurvey.NewRootSurveyPacket(seg)
+	if err != nil {
+		panic(err)
+	}
+	response, err := root.NewResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	response.SetNamespace(ns)
+
+	if cab, ok := peerstore.GetCertifiedAddrBook(surv.h.Peerstore()); ok {
+		env := cab.GetPeerRecord(surv.h.ID())
+		rec, err := env.Marshal()
+		if err != nil {
+			return nil, err
+		}
+		response.SetEnvelope(rec)
+	}
+	return root.Message().MarshalPacked()
+}
+
+func (surv *Survey) handleResponse(response cpSurvey.SurveyResponse) {
+	var (
+		finder chan peer.AddrInfo
+		rec    peer.PeerRecord
+		err    error
+		ok     bool
+	)
+
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	ns, err := response.Namespace()
+	if err != nil {
+		return
+	}
+
+	if finder, ok = surv.mustFind[ns]; !ok {
+		return
+	}
+
+	envelope, err := response.Envelope()
+	if err != nil {
+		return
+	}
+
+	if _, err = record.ConsumeTypedEnvelope(envelope, &rec); err != nil {
+		return
+	}
+	select {
+	case finder <- peer.AddrInfo{ID: rec.PeerID, Addrs: rec.Addrs}:
+	default:
+		return
+	}
+}
+
+func (surv *Survey) Advertise(ctx context.Context, ns string, opt ...discovery.Option) (time.Duration, error) {
+	if err := surv.err.Load(); err != nil {
```

Comment:
The state of `surv.err` may change at any point, so we'll need some additional synchronization.